### PR TITLE
Add CI-build script

### PR DESCRIPTION
**Pull request by gunnarx**
Mostly untested, but it starts a build at least.

Proposal for discussion.  Likely needs a couple of updates before merging.

One concern is where the dependent layer are to be placed, but the current expectation seems to be that poky/ meta-openembeded/ and meta-gplv2/ should be adjacent to the meta-ivi "root" (repository working dir):

```
$ cd myworkdir
$ tree -L 2 -d
├── meta-gplv2
│   ├── ... (abbreviated)
├── meta-ivi
│   ├── build
│   ├── meta-ivi
│   ├── meta-ivi-bsp
│   ├── meta-ivi-test
│   └── scripts
├── meta-openembedded
│   ├── ... (abbreviated)
└── poky
│   ├── ... (abbreviated)
```

The script was tested as follows:

```
$ meta-ivi/scripts/ci-build.sh
```
or
```
$ cd meta-ivi
$ scripts/ci-build.sh
```
Both seem to work as they should.

### CODE REVIEW

**Comment by gunnarx** [comment]
I believe it works now.  Please review since there might be some old cruft left from GDP version it was based on.  
- If reading through it makes sense and fix what you can see
- If it builds the image correctly
... then it can be merged

**Comment by slawr** [comment]
I need to leave early today but I'll try to kick off a local build for R-Car either today or tomorrow and will let you know the result.

File: scripts/ci-build.sh
Comment (jeremiah) [inline scripts/ci-build.sh]: Why not just call ${other_variable}?

File: scripts/init.sh
Comment (jeremiah) [inline scripts/init.sh]: I see this is just a comment but it says "miranda" when other scripts say "pulsar". Does this matter? Will it confuse someone (like it confuses me. :)

File: scripts/ci-build.sh
Comment (gunnarx) [reply scripts/ci-build.sh]: ${var} returns the value stored in var
deref var returns the value stored in the variable whose name is stored in var
e.g.
```
x=y
y=z
echo $x    <- prints out "y"
echo $(deref x)    <- prints out "z"
```

Comment (gunnarx) [reply scripts/ci-build.sh]: also
```
x=y
y=z
echo ${$x}  <- you'd think this should work, but...
--> bash: ${$x}: bad substitution
```
But maybe there's another way...

File: scripts/init.sh
Comment (gunnarx) [reply scripts/init.sh]: This (and all other files) should not be part of PR - only ci-build.sh was intended.
I must have goofed on the last commit.  I'll fix and update.

File: scripts/ci-build.sh
Comment (gunnarx) [inline scripts/ci-build.sh]: Could more succinctly use string substitution directly in bash, it's just that I'm familiar with the sed syntax...

File: scripts/ci-build.sh
Comment (gunnarx) [inline scripts/ci-build.sh:141]: Could use string substitution directly in bash, it's just that I'm familiar with the sed syntax...

**Comment by gunnarx** [comment]
> This (and all other files) should not be part of PR - only ci-build.sh was intended. I must have goofed on the last commit. I'll fix and update.

The image definition files are now gone.

**Comment by gunnarx** [comment]
Build script seems to work now, according to [this test pipeline](https://go.genivi.org/go/tab/build/detail/meta-ivi-with-script/13/Build/1/init_and_bitbake).  (A few comments were removed after that, but it should be the same script)

Might be some corner-case bugs remaining of course. 

Some instructions to copy artifacts to a staging/ directory should be added but I'll let @frznlogic consider how to add that later.  

I recommend merging now.

**Comment by frznlogic** [comment]
Hi, just saw that you mentioned it should be pulled in now.

Some general comments/questions. 
Wouldn't it be better to break the scripts into their own repository for code reusability and to avoid duplication of efforts? Ie, a separate repository containing this script modified to be more general,  provide configurations within the repo or the go.cd environment for what to fetch etc. 

The bash code is a bit under-documented, some comments on the intention of the functions would be nice. Please explain the trickery and the rationale instead of calling it trickery also :)

**Comment by frznlogic** [comment]
Also, please remove the WIP in the PR name :)

**Comment by gunnarx** [comment]
> Wouldn't it be better to break the scripts into their own repository for code reusability and to avoid duplication of efforts? 

You mean the script used by meta-ivi and GDP?  Yes, there is a bit of duplication between them, but I felt the argument to keep them with the source (below) stronger so far.  I am also unsure how different GDP & meta-ivi scripts will become, or continue to be.  (Related to the strategy of parent-project for meta-ivi, with submodules, etc.) so this is a start, with potential for refactoring later.

> Ie, a separate repository containing this script modified to be more general, 

Basically my argument against this has been that ci-build.sh *can* be usable also for advanced users (and I hope we can take any steps to improve that usability, such as documentation).   For me that points towards keeping the build-script with the source that users already have, instead of requiring users to pick up another new repository.  We also used to have that, a "baseline-ci-helper" repo for Go builds, but it was just overhead...

There are a few dependencies between script & source also (for example the list of sub-layers inside the script).  When there's a dependency that needs to be in sync with eachother, it points towards keeping it in one repo if possible.  There's the possibility of a "scripts" submodule perhaps, but I don't know if we are there yet.

> provide configurations within the repo or the go.cd environment for what to fetch etc.

I find I often prefer a straight-forward approach, using git branches and such, to manage differences, instead of creating too much structure.  Unless the problem/project is big enough to actually gain speed in the long run by creating it.  For this you need to decide on some "configuration language" of sorts, then keep those configurations in sync with each project, and so on.
I'm not sure this build stuff big enough  to warrant the effort.  BUT, if you envision something larger here, like reuse in pelux as well then I might be convinced I guess. I'm open to suggestions.

> The bash code is a bit under-documented, some comments on the intention of the functions would be nice. 

Agreed, but could it be future improvement?  (I was hoping you take over the script, assuming it gets merged here...).   If you review-comment on the lines that you feel must be fixed, I'll address those.
**EDIT** Added comments in bf5e8d3

> Please explain the trickery and the rationale instead of calling it trickery also :)

In my opinion, the comment just above the deref function explains what the function does as clearly as can be done.  My answer to Jeremiah basically repeated the comment. The need for the function can be seen where it is used, but OK, I'll add rationale.

**Comment by gunnarx** [comment]
@frznlogic Please review latest commit for more documentation.  We could squash before merge if you prefer.

File: scripts/ci-build.sh
Comment (gunnarx) [reply scripts/ci-build.sh]: The reason something like it is needed here is that the variable *name* itself is generated by the program (it's not known, like in the above example of x & y).

**Comment by slawr** [comment]
>>     Ie, a separate repository containing this script modified to be more general,
> 
> Basically my argument against this has been that ci-build.sh can be usable also for advanced users (and I hope we can take any steps to improve that usability, such as documentation). For me that points towards keeping the build-script with the source that users already have, instead of requiring users to pick up another new repository.

If its simple to use I could see that having it in the source may help contribution, including component maintainers, if it simplifies setup of the baseline when they are not used to doing it.

**Comment by gunnarx** [comment]
Well, yes the script is simple(r) to use than the alternatives.  Currently the script checks out the right commits (taken from README), but does *not* clone the layers themselves.  I think that should be added, unless we have some other parent-project setup.

So as of now: these are basically the minimal steps you must do:
```bash
$ git clone git://github.com/GENIVI/meta-ivi
$ git clone git://git.yoctoproject.org/poky
$ git clone git://git.openembedded.org/meta-openembedded
$ git clone git://git.yoctoproject.org/meta-gplv2
$ export TARGET=qemuarm64
$ meta-ivi/scripts/ci-build.sh
```
That should start the build.  It still takes some steps away.  Steps 2-4 ought to be eliminated in some way - please let me know how you think it should be done.

**Comment by slawr** [comment]
Thanks for the note and sorry for not trying the script earlier like I said I would. It slipped my mind with all the Domain Interaction work recently. The note makes it clear where I would do the extras for a community board.

When I think of baseline builds I see two broad use cases, building the "as is" baseline and the baseline as part of something larger. The former could be for the official qemu or community targets. Community targets might mean the addition of a layer or two but you are not radically reshaping the distro. For me they are the principal targets of such "helpers". It may simply be that the helper needs to be flexible enough to support the community targets either by extension or manual local work to add things.

Personally I would not worry too much about the second use case, baseline as part of something much larger, currently. Such a distro is likely to have its own tooling anyway.

File: scripts/ci-build.sh
Comment (frznlogic) [inline scripts/ci-build.sh]: I'd prefer the default for DL_DIR to be false

File: scripts/ci-build.sh
Comment (frznlogic) [inline scripts/ci-build.sh]: I'd prefer the default for SSTATE_DIR to be false. 

It's easy to set this for the CI build and generates a fairly complex error message for whatever user right now.

File: scripts/ci-build.sh
Comment (frznlogic) [inline scripts/ci-build.sh:12]: All of these paths are specific to the CI environment on go.genivi.org. I'd prefer them to be either be tied to $HOME or $PWD if possible and then change them locally in the CI environment.

File: scripts/ci-build.sh
Comment (frznlogic) [inline scripts/ci-build.sh]: The script will continue appending to an existing local.conf file for every time you try to run it. Ie, we wind up with the MIRRORS_append once every time we try to run this script as it is now.

File: scripts/ci-build.sh
Comment (frznlogic) [inline scripts/ci-build.sh]: For some reason we wind up building in <path>/meta-ivi/build/... I'd prefer to keep the build directory in <path>/build, the same level as the other meta-layers.

**Review by frznlogic — CHANGES_REQUESTED** [review]
Looks better with the comments etc. I've tried running the script and had some more minor issues with the defaults basically. What it boils down to imho, it would be nicer to move the burden of handling defaults from the users to the CI environment.

File: scripts/ci-build.sh
Comment (gunnarx) [reply scripts/ci-build.sh:141]: ... patches welcome :)

File: scripts/ci-build.sh
Comment (gunnarx) [reply scripts/ci-build.sh]: OK, sure, that could be changed.  This is a remnant from GDP where build/ and scripts/ are in parallel (since GDP git is a parent project but meta-ivi is just a layer).

File: scripts/ci-build.sh
Comment (gunnarx) [reply scripts/ci-build.sh]: OK

File: scripts/ci-build.sh
Comment (gunnarx) [reply scripts/ci-build.sh]: OK

File: scripts/ci-build.sh
Comment (gunnarx) [inline scripts/ci-build.sh:508]: By the way, there isn't very deep thought behind the selection of move vs copy here.  Mostly I guess the idea should be to move large things instead of copy, to save disk space and time but to copy things like logs that are expected to be found when poking around the build tree...

All of this is sort of optimized for CI and we might want to put a condition on it.

File: scripts/ci-build.sh
Comment (gunnarx) [inline scripts/ci-build.sh]: Note that once the default for "Use standard dirs" is false, the user will be able to specify their own values for SGX bin, DL_DIR an SSTATE_DIR.

define_with_default() will not change a variable if it already has a value in environment.

File: scripts/ci-build.sh
Comment (gunnarx) [reply scripts/ci-build.sh:12]: These have the prefix AGENT_ because they are for the build agents.  They affect the user build only depending on REUSE_*, and if the user does not set its own SGX_DRIVERS location, etc.
So I propose to keep them.  We could set a user-friendly default, but what would that be?
Proposal?  $HOME/something?

File: scripts/ci-build.sh
Comment (gunnarx) [reply scripts/ci-build.sh]: Agreed, this is because CI environments are thrown away frequently.  So this could be improved for end-user, yes.

**Comment by gunnarx** [comment]
Squashed one commit that was superfluous.  Question now is if we squash more or keep this logical history.
@frznlogic I'm not sure if your review is still valid or can be approved now?  I think you need to try the script and decide if it's OK.

**Review by frznlogic — APPROVED** [review]

**Comment by frznlogic** [comment]
@gunnarx I'm good with it, let's get this in imho. Waiting for the builds to finish and then pulling it in :)

**Pull request merged**
